### PR TITLE
Relative Skins (+Attachments)

### PR DESCRIPTION
**Pull request by CJCrafter**
The old skin system code is quite statically typed. This PR adds an interface for both skins and skin selectors. This allows us to add relative skins that simplify the config, especially when using skins and attachments to modify the weapon skin.

### CODE REVIEW

**Comment by CJCrafter** [comment]
The new resource pack and config will break all users who still use an old version of the plugin.

**Review by DeeCaaD — CHANGES_REQUESTED** [review]
1) Because of the relative skins, you'll probably have to make serializers check that any combination of weapon A skins does not overlap with any combination of weapon B skins. This is pretty big QOL thing, which is easier to implement as code rather than having confused users

2) CS + CSP converter to support new format

File: WeaponMechanics/src/main/java/me/deecaad/weaponmechanics/weapon/skin/RelativeSkinSelector.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+        // Check for attachments
+        ConfigurationSection attachmentSection = data.of("Attachments").assertType(ConfigurationSection.class).get(null);
+        Set<String> attachmentKeys = attachmentSection == null ? null : attachmentSection.getKeys(false);
+
+        if (attachmentKeys != null) {
+            for (String attachment : attachmentKeys) {
+
+                // We can't really check if an attachment exists since we
+                // cannot access the WMP API, so users may be prone to error
```
[the remark this answers is not in the document]
Comment (CJCrafter) [reply WeaponMechanics/src/main/java/me/deecaad/weaponmechanics/weapon/skin/RelativeSkinSelector.java:134]: In the future, a WMP manual check is possible.

[the remark this answers is not in the document]
Comment (CJCrafter) [reply WeaponMechanics/src/main/java/me/deecaad/weaponmechanics/weapon/skin/RelativeSkinSelector.java:134]: No need to update CS+CSP converter, old format still works (+new format is so comparatively "weird" that a perfect converter is not possible).

But adding a "strict format" mode for relative skins is probably a good idea. Basically I have this formula that I like which allows 999 custom weapons, each with 10 options for reloading/scoping/scope stacking/sprinting skins, each with 10 options for WMC `/skin` skins, each with room for 5 attachments "slots" (each attachment slot can have 10 mutually exclusive attachments, so 10 different scopes, for example).

Forcing players to use this format (but being able to turn it off in `config.yml` is a very good idea). A youtube video explaining relative skins in better detail is definitely coming soon.

**Comment by CJCrafter** [comment]
https://github.com/WeaponMechanics/MechanicsMain/pull/341/files?file-filters%5B%5D=.yml&show-deleted-files=true&show-viewed-files=true

^ link to view config changes

File: WeaponMechanics/src/main/java/me/deecaad/weaponmechanics/weapon/skin/RelativeSkinSelector.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (CJCrafter) [reply WeaponMechanics/src/main/java/me/deecaad/weaponmechanics/weapon/skin/RelativeSkinSelector.java:153]: Added: https://github.com/WeaponMechanics/MechanicsMain/pull/341/commits/dc3f4bd6781f4fc2b77aff4eae5ee671da88142c#diff-9f9758eb4521c500c72c11dc6dad08a8cae4ea547e645c9c1f7b9da2215659f9R54-R121

**Pull request merged**